### PR TITLE
Document BuildInfo

### LINKED ISSUE #2106: Support for `BuildInfo`
**Is your feature request related to a problem? Please describe.**
Discussed in https://github.com/VirtusLab/scala-cli/discussions/2051

<div type='discussions-op-text'>

<sup>Originally posted by **ckipp01** April 23, 2023</sup>
I have a use case where I'm building a CLI app. A lot of times with apps like this you want to make sure you can display the version of the app via a command, and many times the value of that version is persisted utilizing a form of `BuildInfo` where based on git, or something else a file is generated and used for compilation. Is there any way to mimic something like this with `scala-cli`? All the solutions I'm coming up with request running `scala-cli` to generate a file first, and then running it again on your main project to utilize that created file. However, it'd be great if this could just be done in a single step somehow. Any ideas?</div>

**Describe the solution you'd like**
We could have a directive/command-line option exposing `BuildInfo` data in the project.
```scala
//> using build.info
```

**Describe alternatives you've considered**
This could potentially also be delivered through code generators. (see #610)

**Additional context**
See the discussions in #2051 and #610.

### PR DESCRIPTION
**Pull request by MaciejG604**
Documentation for #2249 (#2106)

### CODE REVIEW

File: modules/generate-reference-doc/src/main/scala/scala/cli/doc/GenerateReferenceDoc.scala [hunk 1]
```diff
@@ -481,6 +482,92 @@ object GenerateReferenceDoc extends CaseApp[InternalDocOptions] {
...
+        |
+        |    scala-cli config power true
+        |:::""".stripMargin
+    )
+
+    b.section(
+      """During the building process Scala CLI collects information about the project's configuration,
+        |both from the console options and `using directives` found in the project's sources.
+        |You can access this information from your code using the `BuildInfo` object, that's automatically generated for your
+        |build when requested.""".stripMargin
```
Comment (tgodzik) [inline modules/generate-reference-doc/src/main/scala/scala/cli/doc/GenerateReferenceDoc.scala]:
```suggestion
        |build on compile when that information changes.""".stripMargin
```

File: modules/generate-reference-doc/src/main/scala/scala/cli/doc/GenerateReferenceDoc.scala [hunk 2]
```diff
@@ -481,6 +482,92 @@ object GenerateReferenceDoc extends CaseApp[InternalDocOptions] {
...
+
+    b.section(
+      """To enable BuildInfo generation pass the `--build-info` option to Scala CLI or use a
+        |`//> using buildInfo` directive.""".stripMargin
+    )
+
+    b.section(
+      """## Usage
+        |
+        |To access the BuildInfo object you need to import it first. It is available in package `scala.cli.build` so use
```
Comment (tgodzik) [inline modules/generate-reference-doc/src/main/scala/scala/cli/doc/GenerateReferenceDoc.scala]:
```suggestion
        |To access the BuildInfo object you need to import it first. It is available in the package `scala.cli.build` so use
```

**Review by tgodzik — CHANGES_REQUESTED** [review]
Minor comments from me, great work!

File: modules/generate-reference-doc/src/main/scala/scala/cli/doc/GenerateReferenceDoc.scala [hunk 3]
```diff
@@ -481,6 +482,92 @@ object GenerateReferenceDoc extends CaseApp[InternalDocOptions] {
...
+        |
+        |    scala-cli config power true
+        |:::""".stripMargin
+    )
+
+    b.section(
+      """During the building process Scala CLI collects information about the project's configuration,
+        |both from the console options and `using directives` found in the project's sources.
+        |You can access this information from your code using the `BuildInfo` object, that's automatically generated for your
+        |build build on compile when that information changes.""".stripMargin
```
Comment (tgodzik) [inline modules/generate-reference-doc/src/main/scala/scala/cli/doc/GenerateReferenceDoc.scala]: Build build -> build

**Review by tgodzik — CHANGES_REQUESTED** [review]

File: website/docs/reference/build-info.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+BuildInfo is a restricted feature and requires setting the `--power` option to be used.
+You can pass it explicitly or set it globally by running:
+
+    scala-cli config power true
+:::
+
+During the building process Scala CLI collects information about the project's configuration,
+both from the console options and `using directives` found in the project's sources.
+You can access this information from your code using the `BuildInfo` object, that's automatically generated for your
+build build on compile when that information changes.
```
Comment (lwronski) [inline website/docs/reference/build-info.md]:
```suggestion
You can access this information from your code using the `BuildInfo` object. That is added to the project's classpath and automatically generated for your build on compile when that information changes.
```

Comment (MaciejG604) [reply website/docs/reference/build-info.md]: I added this in the beginning of the next section. WDYT?

Comment (lwronski) [reply website/docs/reference/build-info.md]: It's fine.

**Review by lwronski — APPROVED** [review]

**Review by tgodzik — APPROVED** [review]
LGTM!

**Pull request merged**